File: src/keybindings.rs
//! Configurable local keybindings for terminal actions.

use crate::input::{Key, KeyEvent, Modifiers};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LocalAction {
    Copy,
    Paste,
    ZoomIn,
    ZoomOut,
    ZoomReset,
}

#[derive(Debug, Clone)]
pub struct KeybindingsConfig {
    pub copy: Vec<KeyBinding>,
    pub paste: Vec<KeyBinding>,
    pub zoom_in: Vec<KeyBinding>,
    pub zoom_out: Vec<KeyBinding>,
    pub zoom_reset: Vec<KeyBinding>,
}

impl Default for KeybindingsConfig {
    fn default() -> Self {
        Self {
            copy: parse_binding_list("ctrl+shift+c").expect("default copy binding must parse"),
            paste: parse_binding_list("ctrl+shift+v").expect("default paste binding must parse"),
            zoom_in: parse_binding_list("ctrl+=, ctrl+plus")
                .expect("default zoom_in binding must parse"),
            zoom_out: parse_binding_list("ctrl+-, ctrl+minus")
                .expect("default zoom_out binding must parse"),
            zoom_reset: parse_binding_list("ctrl+0")
                .expect("default zoom_reset binding must parse"),
        }
    }
}

impl KeybindingsConfig {
    pub fn matches(&self, action: LocalAction, event: &KeyEvent, modifiers: Modifiers) -> bool {
        let bindings = match action {
            LocalAction::Copy => &self.copy,
            LocalAction::Paste => &self.paste,
            LocalAction::ZoomIn => &self.zoom_in,
            LocalAction::ZoomOut => &self.zoom_out,
            LocalAction::ZoomReset => &self.zoom_reset,
        };

        bindings
            .iter()
            .any(|binding| binding.matches(event, modifiers))
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct KeyBinding {
    modifiers: BindingModifiers,
    key: BindingKey,
}

impl KeyBinding {
    pub fn parse(spec: &str) -> Result<Self, String> {
        let spec = spec.trim();
        if spec.is_empty() {
            return Err("empty keybinding".to_string());
        }

        let parts: Vec<_> = spec.split('+').map(str::trim).collect();
        if parts.is_empty() {
            return Err("empty keybinding".to_string());
        }

        let key_part = parts.last().copied().unwrap_or_default();
        if key_part.is_empty() {
            return Err(format!(
                "missing key in binding '{spec}' (use names like 'plus' for '+')"
            ));
        }

        let mut modifiers = BindingModifiers::default();
        for modifier in &parts[..parts.len().saturating_sub(1)] {
            if modifier.is_empty() {
                return Err(format!(
                    "invalid modifier in binding '{spec}' (use names like 'plus' for '+')"
                ));
            }
            modifiers.apply(modifier)?;
        }

        Ok(Self {
            modifiers,
            key: BindingKey::parse(key_part)?,
        })
    }

    pub fn matches(&self, event: &KeyEvent, modifiers: Modifiers) -> bool {
        event.pressed
            && self.key.matches(&event.key)
            && self
                .modifiers
                .matches(modifiers, self.key.allows_implicit_shift())
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
struct BindingModifiers {
    ctrl: bool,
    shift: bool,
    alt: bool,
}

impl BindingModifiers {
    fn apply(&mut self, modifier: &str) -> Result<(), String> {
        match modifier.to_ascii_lowercase().as_str() {
            "ctrl" | "control" => self.ctrl = true,
            "shift" => self.shift = true,
            "alt" => self.alt = true,
            other => return Err(format!("unknown modifier '{other}'")),
        }
        Ok(())
    }

    fn matches(&self, modifiers: Modifiers, allow_extra_shift: bool) -> bool {
        self.ctrl == modifiers.ctrl
            && self.alt == modifiers.alt
            && (self.shift == modifiers.shift
                || (allow_extra_shift && !self.shift && modifiers.shift))
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
enum BindingKey {
    Text(String),
    Named(NamedKey),
}

impl BindingKey {
    fn parse(token: &str) -> Result<Self, String> {
        let token = token.trim();
        let lower = token.to_ascii_lowercase();

        let named = match lower.as_str() {
            "escape" | "esc" => Some(NamedKey::Escape),
            "enter" | "return" => Some(NamedKey::Enter),
            "backspace" => Some(NamedKey::Backspace),
            "tab" => Some(NamedKey::Tab),
            "space" => Some(NamedKey::Space),
            "up" | "arrowup" => Some(NamedKey::ArrowUp),
            "down" | "arrowdown" => Some(NamedKey::ArrowDown),
            "left" | "arrowleft" => Some(NamedKey::ArrowLeft),
            "right" | "arrowright" => Some(NamedKey::ArrowRight),
            "home" => Some(NamedKey::Home),
            "end" => Some(NamedKey::End),
            "pageup" | "pgup" => Some(NamedKey::PageUp),
            "pagedown" | "pgdown" => Some(NamedKey::PageDown),
            "delete" | "del" => Some(NamedKey::Delete),
            "insert" | "ins" => Some(NamedKey::Insert),
            _ => None,
        };
        if let Some(named) = named {
            return Ok(Self::Named(named));
        }

        if let Some(function) = parse_function_key(&lower) {
            return Ok(Self::Named(NamedKey::Function(function)));
        }

        let text = match lower.as_str() {
            "plus" => "+",
            "minus" => "-",
            "equal" | "equals" => "=",
            "comma" => ",",
            "period" | "dot" => ".",
            "slash" => "/",
            "backslash" => "\\",
            "semicolon" => ";",
            "apostrophe" | "quote" => "'",
            "grave" | "backtick" => "`",
            "leftbracket" | "lbracket" => "[",
            "rightbracket" | "rbracket" => "]",
            _ => token,
        };

        if text.chars().count() == 1 {
            return Ok(Self::Text(normalize_text_key(text)));
        }

        Err(format!("unknown key '{token}'"))
    }

    fn allows_implicit_shift(&self) -> bool {
        matches!(self, Self::Text(text) if text.chars().count() == 1
            && !text.chars().next().unwrap_or_default().is_ascii_alphabetic())
    }

    fn matches(&self, key: &Key) -> bool {
        match (self, key) {
            (Self::Text(expected), Key::Text(actual)) => normalize_text_key(actual) == *expected,
            (Self::Named(NamedKey::Escape), Key::Escape)
            | (Self::Named(NamedKey::Enter), Key::Enter)
            | (Self::Named(NamedKey::Backspace), Key::Backspace)
            | (Self::Named(NamedKey::Tab), Key::Tab)
            | (Self::Named(NamedKey::Space), Key::Space)
            | (Self::Named(NamedKey::ArrowUp), Key::ArrowUp)
            | (Self::Named(NamedKey::ArrowDown), Key::ArrowDown)
            | (Self::Named(NamedKey::ArrowLeft), Key::ArrowLeft)
            | (Self::Named(NamedKey::ArrowRight), Key::ArrowRight)
            | (Self::Named(NamedKey::Home), Key::Home)
            | (Self::Named(NamedKey::End), Key::End)
            | (Self::Named(NamedKey::PageUp), Key::PageUp)
            | (Self::Named(NamedKey::PageDown), Key::PageDown)
            | (Self::Named(NamedKey::Delete), Key::Delete)
            | (Self::Named(NamedKey::Insert), Key::Insert) => true,
            (Self::Named(NamedKey::Function(expected)), Key::F(actual)) => expected == actual,
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum NamedKey {
    Escape,
    Enter,
    Backspace,
    Tab,
    Space,
    ArrowUp,
    ArrowDown,
    ArrowLeft,
    ArrowRight,
    Home,
    End,
    PageUp,
    PageDown,
    Delete,
    Insert,
    Function(u8),
}

fn parse_function_key(token: &str) -> Option<u8> {
    let suffix = token.strip_prefix('f')?;
    let num = suffix.parse::<u8>().ok()?;
    (1..=12).contains(&num).then_some(num)
}

fn normalize_text_key(text: &str) -> String {
    let mut chars = text.chars();
    match (chars.next(), chars.next()) {
        (Some(ch), None) if ch.is_ascii_alphabetic() => ch.to_ascii_lowercase().to_string(),
        _ => text.to_string(),
    }
}

pub fn parse_binding_list(spec: &str) -> Result<Vec<KeyBinding>, String> {
    let mut bindings = Vec::new();
    for binding in spec.split(',').map(str::trim).filter(|s| !s.is_empty()) {
        bindings.push(KeyBinding::parse(binding)?);
    }

    if bindings.is_empty() {
        return Err("expected at least one keybinding".to_string());
    }

    Ok(bindings)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn mods(ctrl: bool, shift: bool, alt: bool) -> Modifiers {
        Modifiers { ctrl, shift, alt }
    }

    #[test]
    fn parses_basic_binding() {
        let binding = KeyBinding::parse("ctrl+shift+c").unwrap();
        assert!(binding.matches(
            &KeyEvent {
                pressed: true,
                key: Key::Text("C".to_string()),
            },
            mods(true, true, false)
        ));
        assert!(!binding.matches(
            &KeyEvent {
                pressed: true,
                key: Key::Text("c".to_string()),
            },
            mods(true, false, false)
        ));
    }

    #[test]
    fn parses_named_symbol_aliases() {
        let plus = KeyBinding::parse("ctrl+plus").unwrap();
        assert!(plus.matches(
            &KeyEvent {
                pressed: true,
                key: Key::Text("+".to_string()),
            },
            mods(true, true, false)
        ));

        let minus = KeyBinding::parse("ctrl+minus").unwrap();
        assert!(minus.matches(
            &KeyEvent {
                pressed: true,
                key: Key::Text("-".to_string()),
            },
            mods(true, false, false)
        ));
    }

    #[test]
    fn parses_multiple_bindings() {
        let bindings = parse_binding_list("ctrl+=, ctrl+plus").unwrap();
        assert_eq!(bindings.len(), 2);
    }

    #[test]
    fn supports_named_keys() {
        let binding = KeyBinding::parse("alt+pageup").unwrap();
        assert!(binding.matches(
            &KeyEvent {
                pressed: true,
                key: Key::PageUp,
            },
            mods(false, false, true)
        ));
    }

    #[test]
    fn rejects_missing_key() {
        assert!(KeyBinding::parse("ctrl+").is_err());
    }

    #[test]
    fn default_config_matches_existing_shortcuts() {
        let bindings = KeybindingsConfig::default();
        assert!(bindings.matches(
            LocalAction::Copy,
            &KeyEvent {
                pressed: true,
                key: Key::Text("c".to_string()),
            },
            mods(true, true, false)
        ));
        assert!(bindings.matches(
            LocalAction::Paste,
            &KeyEvent {
                pressed: true,
                key: Key::Text("v".to_string()),
            },
            mods(true, true, false)
        ));
        assert!(bindings.matches(
            LocalAction::ZoomIn,
            &KeyEvent {
                pressed: true,
                key: Key::Text("=".to_string()),
            },
            mods(true, false, false)
        ));
        assert!(bindings.matches(
            LocalAction::ZoomIn,
            &KeyEvent {
                pressed: true,
                key: Key::Text("+".to_string()),
            },
            mods(true, true, false)
        ));
    }
}

File: src/keybindings.c
#include "rudo/keybindings.h"

#include <ctype.h>
#include <stdio.h>

static void set_err(char **err, const char *msg) { if (err) *err = rudo_strdup(msg); }
static void set_errf(char **err, const char *fmt, const char *arg) { if (!err) return; size_t n = snprintf(NULL, 0, fmt, arg); *err = rudo_malloc(n + 1); snprintf(*err, n + 1, fmt, arg); }

static char *lower_dup(const char *s) {
    size_t i, n = strlen(s);
    char *o = rudo_malloc(n + 1);
    for (i = 0; i < n; ++i) o[i] = (char)tolower((unsigned char)s[i]);
    o[n] = 0;
    return o;
}

static void normalize_text_inplace(char *s) {
    if (s[0] && !s[1] && isalpha((unsigned char)s[0])) s[0] = (char)tolower((unsigned char)s[0]);
}

void rudo_keybinding_destroy(rudo_keybinding *binding) { rudo_key_destroy(&binding->key); }
void rudo_keybinding_list_init(rudo_keybinding_list *list) { list->v = NULL; list->len = list->cap = 0; }

void rudo_keybinding_list_destroy(rudo_keybinding_list *list) {
    size_t i;
    for (i = 0; i < list->len; ++i) rudo_keybinding_destroy(&list->v[i]);
    free(list->v);
    list->v = NULL; list->len = list->cap = 0;
}

bool rudo_keybinding_list_push(rudo_keybinding_list *list, const rudo_keybinding *binding) {
    if (list->len == list->cap) {
        size_t cap = list->cap ? list->cap * 2 : 4;
        list->v = rudo_realloc(list->v, cap * sizeof(*list->v));
        list->cap = cap;
    }
    list->v[list->len].modifiers = binding->modifiers;
    rudo_key_copy(&list->v[list->len].key, &binding->key);
    list->len++;
    return true;
}

static bool parse_named_or_text(rudo_key *key, const char *token, char **err) {
    char *lower = lower_dup(token);
    if (rudo_streq(lower, "escape") || rudo_streq(lower, "esc")) rudo_key_init_named(key, RUDO_KEY_ESCAPE);
    else if (rudo_streq(lower, "enter") || rudo_streq(lower, "return")) rudo_key_init_named(key, RUDO_KEY_ENTER);
    else if (rudo_streq(lower, "backspace")) rudo_key_init_named(key, RUDO_KEY_BACKSPACE);
    else if (rudo_streq(lower, "tab")) rudo_key_init_named(key, RUDO_KEY_TAB);
    else if (rudo_streq(lower, "space")) rudo_key_init_named(key, RUDO_KEY_SPACE);
    else if (rudo_streq(lower, "up") || rudo_streq(lower, "arrowup")) rudo_key_init_named(key, RUDO_KEY_ARROW_UP);
    else if (rudo_streq(lower, "down") || rudo_streq(lower, "arrowdown")) rudo_key_init_named(key, RUDO_KEY_ARROW_DOWN);
    else if (rudo_streq(lower, "left") || rudo_streq(lower, "arrowleft")) rudo_key_init_named(key, RUDO_KEY_ARROW_LEFT);
    else if (rudo_streq(lower, "right") || rudo_streq(lower, "arrowright")) rudo_key_init_named(key, RUDO_KEY_ARROW_RIGHT);
    else if (rudo_streq(lower, "home")) rudo_key_init_named(key, RUDO_KEY_HOME);
    else if (rudo_streq(lower, "end")) rudo_key_init_named(key, RUDO_KEY_END);
    else if (rudo_streq(lower, "pageup") || rudo_streq(lower, "pgup")) rudo_key_init_named(key, RUDO_KEY_PAGE_UP);
    else if (rudo_streq(lower, "pagedown") || rudo_streq(lower, "pgdown")) rudo_key_init_named(key, RUDO_KEY_PAGE_DOWN);
    else if (rudo_streq(lower, "delete") || rudo_streq(lower, "del")) rudo_key_init_named(key, RUDO_KEY_DELETE);
    else if (rudo_streq(lower, "insert") || rudo_streq(lower, "ins")) rudo_key_init_named(key, RUDO_KEY_INSERT);
    else if (lower[0] == 'f' && lower[1]) {
        char *end = NULL; long n = strtol(lower + 1, &end, 10);
        if (*end == 0 && n >= 1 && n <= 12) rudo_key_init_function(key, (uint8_t)n); else goto text_fallback;
    } else {
text_fallback:
        {
            const char *text = NULL;
            if (rudo_streq(lower, "plus")) text = "+";
            else if (rudo_streq(lower, "minus")) text = "-";
            else if (rudo_streq(lower, "equal") || rudo_streq(lower, "equals")) text = "=";
            else if (rudo_streq(lower, "comma")) text = ",";
            else if (rudo_streq(lower, "period") || rudo_streq(lower, "dot")) text = ".";
            else if (rudo_streq(lower, "slash")) text = "/";
            else if (rudo_streq(lower, "backslash")) text = "\\";
            else if (rudo_streq(lower, "semicolon")) text = ";";
            else if (rudo_streq(lower, "apostrophe") || rudo_streq(lower, "quote")) text = "'";
            else if (rudo_streq(lower, "grave") || rudo_streq(lower, "backtick")) text = "`";
            else if (rudo_streq(lower, "leftbracket") || rudo_streq(lower, "lbracket")) text = "[";
            else if (rudo_streq(lower, "rightbracket") || rudo_streq(lower, "rbracket")) text = "]";
            else text = token;
            if (strlen(text) != 1) { set_errf(err, "unknown key '%s'", token); free(lower); return false; }
            rudo_key_init_text(key, text);
            normalize_text_inplace(key->text);
        }
    }
    free(lower);
    return true;
}

bool rudo_keybinding_parse(rudo_keybinding *out, const char *spec, char **err_msg) {
    char *copy, *save = NULL, *part;
    char *segments[16];
    size_t nseg = 0, i;
    if (err_msg) *err_msg = NULL;
    memset(out, 0, sizeof(*out));
    copy = rudo_strdup(spec ? spec : "");
    char *trimmed = rudo_trim(copy);
    if (!*trimmed) { free(copy); set_err(err_msg, "empty keybinding"); return false; }
    for (part = strtok_r(trimmed, "+", &save); part; part = strtok_r(NULL, "+", &save)) segments[nseg++] = rudo_trim(part);
    if (!nseg || !segments[nseg - 1][0]) { free(copy); set_errf(err_msg, "missing key in binding '%s' (use names like 'plus' for '+')", spec); return false; }
    for (i = 0; i + 1 < nseg; ++i) {
        if (!segments[i][0]) { free(copy); set_errf(err_msg, "invalid modifier in binding '%s' (use names like 'plus' for '+')", spec); return false; }
        if (rudo_streq_nocase(segments[i], "ctrl") || rudo_streq_nocase(segments[i], "control")) out->modifiers.ctrl = true;
        else if (rudo_streq_nocase(segments[i], "shift")) out->modifiers.shift = true;
        else if (rudo_streq_nocase(segments[i], "alt")) out->modifiers.alt = true;
        else { free(copy); set_errf(err_msg, "unknown modifier '%s'", segments[i]); return false; }
    }
    if (!parse_named_or_text(&out->key, segments[nseg - 1], err_msg)) { free(copy); return false; }
    free(copy);
    return true;
}

static bool allows_implicit_shift(const rudo_key *key) {
    return key->kind == RUDO_KEY_TEXT && key->text && key->text[0] && !key->text[1] && !isalpha((unsigned char)key->text[0]);
}

bool rudo_keybinding_matches(const rudo_keybinding *binding, const rudo_key_event *event, rudo_modifiers modifiers) {
    if (!event->pressed) return false;
    if (!rudo_key_equal(&binding->key, &event->key)) {
        if (binding->key.kind == RUDO_KEY_TEXT && event->key.kind == RUDO_KEY_TEXT && binding->key.text && event->key.text) {
            char *a = rudo_strdup(binding->key.text), *b = rudo_strdup(event->key.text);
            normalize_text_inplace(a); normalize_text_inplace(b);
            bool ok = rudo_streq(a, b);
            free(a); free(b);
            if (!ok) return false;
        } else return false;
    }
    if (binding->modifiers.ctrl != modifiers.ctrl) return false;
    if (binding->modifiers.alt != modifiers.alt) return false;
    if (binding->modifiers.shift == modifiers.shift) return true;
    return allows_implicit_shift(&binding->key) && !binding->modifiers.shift && modifiers.shift;
}

bool rudo_parse_binding_list(rudo_keybinding_list *out, const char *spec, char **err_msg) {
    char *copy, *save = NULL, *part;
    bool any = false;
    rudo_keybinding tmp;
    rudo_keybinding_list_init(out);
    if (err_msg) *err_msg = NULL;
    copy = rudo_strdup(spec ? spec : "");
    for (part = strtok_r(copy, ",", &save); part; part = strtok_r(NULL, ",", &save)) {
        char *trimmed = rudo_trim(part);
        char *err = NULL;
        if (!*trimmed) continue;
        any = true;
        if (!rudo_keybinding_parse(&tmp, trimmed, &err)) { if (err_msg) *err_msg = err; else free(err); free(copy); rudo_keybinding_list_destroy(out); return false; }
        rudo_keybinding_list_push(out, &tmp);
        rudo_keybinding_destroy(&tmp);
    }
    free(copy);
    if (!any) { set_err(err_msg, "expected at least one keybinding"); return false; }
    return true;
}

static void init_action_defaults(rudo_keybinding_list *list, const char *spec) {
    char *err = NULL;
    if (!rudo_parse_binding_list(list, spec, &err)) { free(err); abort(); }
}

void rudo_keybindings_config_init_default(rudo_keybindings_config *cfg) {
    init_action_defaults(&cfg->copy, "ctrl+shift+c");
    init_action_defaults(&cfg->paste, "ctrl+shift+v");
    init_action_defaults(&cfg->zoom_in, "ctrl+equal, ctrl+plus");
    init_action_defaults(&cfg->zoom_out, "ctrl+minus");
    init_action_defaults(&cfg->zoom_reset, "ctrl+0");
}

void rudo_keybindings_config_destroy(rudo_keybindings_config *cfg) {
    rudo_keybinding_list_destroy(&cfg->copy);
    rudo_keybinding_list_destroy(&cfg->paste);
    rudo_keybinding_list_destroy(&cfg->zoom_in);
    rudo_keybinding_list_destroy(&cfg->zoom_out);
    rudo_keybinding_list_destroy(&cfg->zoom_reset);
}

rudo_keybinding_list *rudo_keybindings_action_list(rudo_keybindings_config *cfg, rudo_local_action action) {
    switch (action) {
        case RUDO_LOCAL_ACTION_COPY: return &cfg->copy;
        case RUDO_LOCAL_ACTION_PASTE: return &cfg->paste;
        case RUDO_LOCAL_ACTION_ZOOM_IN: return &cfg->zoom_in;
        case RUDO_LOCAL_ACTION_ZOOM_OUT: return &cfg->zoom_out;
        case RUDO_LOCAL_ACTION_ZOOM_RESET: return &cfg->zoom_reset;
        default: return NULL;
    }
}

const rudo_keybinding_list *rudo_keybindings_action_list_const(const rudo_keybindings_config *cfg, rudo_local_action action) {
    return (const rudo_keybinding_list *)rudo_keybindings_action_list((rudo_keybindings_config *)cfg, action);
}

bool rudo_keybindings_matches(const rudo_keybindings_config *cfg, rudo_local_action action, const rudo_key_event *event, rudo_modifiers modifiers) {
    size_t i;
    const rudo_keybinding_list *list = rudo_keybindings_action_list_const(cfg, action);
    if (!list) return false;
    for (i = 0; i < list->len; ++i) if (rudo_keybinding_matches(&list->v[i], event, modifiers)) return true;
    return false;
}
